lib/fast-path.ts

```diff
@@ -438,60 +433,59 @@ FPp.needsParens = function(assumeExpressionContext) {
 
     case "CallExpression":
     case "NewExpression":
-      return name === "callee"
-        && parent.callee === node;
+      return parent.callee === node;
 
     case "ConditionalExpression":
-      return name === "test"
-        && parent.test === node;
+      return parent.test === node;
 
     case "MemberExpression":
-      return name === "object"
-        && parent.object === node;
+      return parent.object === node;
 
     default:
       return false;
     }
 
   case "ArrowFunctionExpression":
-    if (n.CallExpression.check(parent) &&
-        name === 'callee') {
-      return true;
-    }
-
-    if (n.MemberExpression.check(parent) &&
-        name === 'object') {
-      return true;
+    switch(parent.type) {
+    case 'CallExpression':
+      return parent.callee === node;
+    case 'MemberExpression':
+      return parent.object === node;
+    case 'TSAsExpression':
+      return parent.expression === node;
+    default:
+      return isBinary(parent);
     }
 
-    return isBinary(parent);
-
   case "ObjectExpression":
-    if (parent.type === "ArrowFunctionExpression" &&
-        name === "body") {
+    if (parent.type === "ArrowFunctionExpression" && parent.body === node) {
       return true;
     }
 
     break;
 
-  case 'TSAsExpression':
-    if (parent.type === 'ArrowFunctionExpression' &&
-      name === 'body' &&
-      node.expression.type === 'ObjectExpression') {
+  case "TSAsExpression":
+    if (
+      parent.type === "ArrowFunctionExpression" &&
+      parent.body === node &&
+      node.expression.type === "ObjectExpression"
+    ) {
       return true;
     }
     break;
 
   case "CallExpression":
-    if (name === "declaration" &&
-        n.ExportDefaultDeclaration.check(parent) &&
-        n.FunctionExpression.check(node.callee)) {
+    if (
+      parent.declaration === node &&
+      n.ExportDefaultDeclaration.check(parent) &&
+      n.FunctionExpression.check(node.callee)
+    ) {
       return true;
     }
+    break;
```

Comment:
There really aught to be lint rules to catch this sort of thing. It's a potential bug if not an actual one.